dm/dm/master/shardddl/optimist.go

```diff
@@ -185,10 +186,45 @@ func (o *Optimist) RemoveMetaData(task string) error {
 		o.lk.RemoveLock(op.ID)
 	}
 
+	o.lk.RemoveDownstreamMeta(task)
 	o.tk.RemoveTableByTask(task)
 
 	// clear meta data in etcd
-	_, err = optimism.DeleteInfosOperationsTablesSchemasByTask(o.cli, task, lockIDSet)
+	_, err = optimism.DeleteInfosOperationsTablesByTask(o.cli, task, lockIDSet)
+	return err
+}
+
+// RemoveMetaDataWithTaskAndSources removes meta data for a specified task and sources
+// NOTE: this function can only be used when the specified task for source is not running.
+func (o *Optimist) RemoveMetaDataWithTaskAndSources(task string, sources ...string) error {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	if o.closed {
+		return terror.ErrMasterOptimistNotStarted.Generate()
+	}
+
+	dropColumns := make(map[string][]string)
+
+	// gets all locks for this task
+	locks := o.lk.FindLocksByTask(task)
+	for _, lock := range locks {
+		// remove table by sources for related lock
+		if lock != nil {
+			cols := lock.TryRemoveTableBySources(sources)
+			dropColumns[lock.ID] = cols
+			o.logger.Debug("the tables removed from the lock", zap.String("task", task), zap.Strings("sources", sources))
+		}
+		if !lock.HasTables() {
```

Comment:
do we need put `!lock.HasTables()` in to `if lock != nil`

Reply:
not null now, 
8e2520e